galley/pkg/config/processing/session.go

```diff
@@ -0,0 +1,277 @@
+	// allowed transitions are: inactive
+	terminating = sessionState("terminating")
+)
+
+type session struct { // nolint:maligned
+	mu sync.Mutex
+
+	id      int32
+	options RuntimeOptions
+	buffer  *event.Buffer
+
+	state sessionState
+
+	// mesh config state
+	meshCfg    *v1alpha1.MeshConfig
+	meshSynced bool
+
+	doneCh chan struct{}
+}
+
+// newSession creates a new config processing session state. It returns the session, as well as a channel
+// that will be closed upon completion of the session activity.
+func newSession(id int32, o RuntimeOptions) (*session, chan struct{}) {
+	s := &session{
+		id:      id,
+		options: o,
+		buffer:  event.NewBuffer(),
+		state:   inactive,
+		doneCh:  make(chan struct{}),
+	}
+
+	s.buffer.Dispatch(o.Processor)
+	return s, s.doneCh
+}
+
+// start the session. This must be called when state == inactive.
+func (s *session) start() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.state != inactive {
+		panic(fmt.Sprintf("invalid state: %s (expecting inactive)", s.state))
+	}
+	s.transitionTo(starting)
+
+	go s.startSources()
+}
+
+func (s *session) startSources() {
+	scope.Debug("session starting sources...")
+	// start sources after relinquishing lock. This avoids deadlocks.
+	for _, src := range s.options.Sources {
+		src.Start()
+	}
+	scope.Debugf("session source start complete")
+
+	// check the state again. During startup we might have received mesh config, or got signaled for stop.
+	var terminate bool
+	s.mu.Lock()
+	switch s.state {
+	case starting:
+		// This is the expected state. Depending on whether we received mesh config or not we can transition to the
+		// buffering, or processing states.
+		s.transitionTo(buffering)
+		if s.meshSynced {
+			s.startProcessing()
+		}
+
+	case terminating:
+		// stop was called during startup. There is nothing we can do, simply exit.
+		terminate = true
+
+	default:
+		panic(fmt.Sprintf("session.start: unexpected state during session startup: %v", s.state))
+	}
+	s.mu.Unlock()
+
+	if terminate {
+		s.terminate()
+	}
+}
+
+func (s *session) stop() {
+	scope.Debug("session.stop()")
+
+	var terminate bool
+	s.mu.Lock()
+	switch s.state {
+	case starting:
+		// set the state to terminating and let the startup code complete startup steps and deal with termination.
+		s.transitionTo(terminating)
+
+	case buffering, processing:
+		s.transitionTo(terminating)
+		terminate = true
+
+	default:
+		scope.Warnf("session.stop: Invalid state: %v", s.state)
+	}
+	s.mu.Unlock()
+	if terminate {
+		s.terminate()
+	}
+}
+
+func (s *session) terminate() {
+	// must be called outside lock.
+	s.mu.Lock()
+	if s.state != terminating {
+		panic(fmt.Sprintf("invalid state: %s (expecting terminating)", s.state))
+	}
+	s.mu.Unlock()
+
+	scope.Debug("session.terminate: stopping buffer...")
+	s.buffer.Stop()
+	scope.Debug("session.terminate: stopping processor...")
+	s.options.Processor.Stop()
+	scope.Debug("session.terminate: stopping sources...")
+	for _, s := range s.options.Sources {
```

Comment:
I think you should stop the sources first, then the buffer, then the processor.

Reply:
I'm assuming that the `Stop` methods are all blocking no? (If they're not, they should be) ... I see your point regarding buffering. My concern is that by writing the teardown this way, you're assuming you have good-behaving resources that can handle being called after they're stopped. That seems fragile to me.

Reply:
Stops are blocking.  I think I am missing your point about good-behaving resources. Call-after-stop is being avoided here. Calling buffer.Stop first implies events to processor will cease before processor.Stop gets called. We can opt to stop sources first, but it means we will be keep processing events from buffer and/or publishing snapshots in the meantime, which maybe redundant.